packages/lesswrong/server/scripts/makeMigrations.ts

```diff
@@ -55,6 +56,234 @@ const schemaFileHeaderTemplate = `-- GENERATED FILE
+  private static concurrentRegex = /\s+CONCURRENTLY\s+/gi;
+
+  private name: string;
+
+  constructor(private index: CustomPgIndex) {
+    super();
+    const {source, options} = index;
+    const name = source.match(CustomIndexNode.nameRegex)?.[4];
+    if (!name) {
+      throw new Error(`Can't parse name for custom index: ${source}`);
+    }
+    this.name = name;
+
+    const target = source.match(CustomIndexNode.targetRegex)?.[2];
+    if (!target) {
+      throw new Error(`Can't parse target for custom index "${name}"`);
+    }
+    const dependency: SchemaDependency = isValidCollectionName(target)
+      ? {type: "collection", name: target}
+      : {type: "view", name: target};
+    this.addDependency(dependency);
+
+    this.dependencies = this.dependencies.concat(options?.dependencies ?? []);
+  }
+
+  getName() {
+    return this.name;
+  }
+
+  getQuery() {
+    return {
+      compile: () => ({
+        sql: this.index.source.trim().replace(CustomIndexNode.concurrentRegex, " "),
+        args: [],
+      }),
+    };
+  }
+}
+
+class FunctionNode extends Node {
+  private static nameRegex = /^\s*CREATE\s+(OR\s+REPLACE\s+)?FUNCTION\s+([a-zA-Z0-9_]+)/i;
+
+  private name: string;
+
+  constructor(private func: PostgresFunction) {
+    super();
+    const name = func.source.match(FunctionNode.nameRegex)?.[2];
+    if (!name) {
+      throw new Error(`Can't find name for function: ${func.source}`);
+    }
+    const overload = func.overload ? `_${func.overload}` : "";
+    this.name = name + overload;
+    this.dependencies = this.dependencies.concat(func.dependencies ?? []);
+  }
+
+  getName() {
+    return this.name;
+  }
+
+  getQuery() {
+    return {
+      compile: () => ({
+        sql: this.func.source.trim(),
+        args: [],
+      }),
+    };
+  }
+}
+
+class ViewNode extends Node {
+  constructor(private view: PostgresView) {
+    super();
+  }
+
+  getName() {
+    return this.view.getName();
+  }
+
+  getQuery() {
+    return {
+      compile: () => ({
+        sql: this.view.getCreateViewQuery().trim(),
+        args: [],
+      }),
+    };
+  }
+}
+
+class Graph {
+  private nodes: Record<string, Node> = {};
+
+  addNode(node: Node) {
+    const name = node.getName();
+    if (this.nodes[name]) {
+      throw new Error(`Duplicate node names: "${name}"`);
+    }
+    this.nodes[name] = node;
+  }
+
+  addNodes(nodes: Node[]) {
+    for (const node of nodes) {
+      this.addNode(node);
+    }
+  }
+
+  getOverallHash() {
+    return md5(Object.values(this.nodes).map((n) => n.getHash()).sort().join());
+  }
+
+  linearize(): Node[] {
+    const stack: Node[] = [];
+    const unvisited = new Set<string>(Object.keys(this.nodes));
+
+    const depthFirstSearch = (nodeName: string) => {
+      const node = this.nodes[nodeName];
+      if (!node) {
+        throw new Error(`Invalid node: "${nodeName}"`);
+      }
+
+      if (!unvisited.has(nodeName)) {
+        return;
+      }
+      unvisited.delete(nodeName);
```

Comment:
Probably doesn't matter in practice, but does this let through circular dependencies?  Seems like it if ran into one, it'd just do an early return on L270 (the same as if it'd run into a dependency that wasn't circular, just shared with something else that had already been visited).